var/spack/repos/builtin/packages/axom/package.py

```diff
@@ -308,13 +308,24 @@ def hostconfig(self, spec, prefix):
         cfg.write("#------------------{0}\n\n".format("-" * 60))
 
         if "+mpi" in spec:
+            mpicc_path = spec['mpi'].mpicc
+            mpicxx_path = spec['mpi'].mpicxx
+            mpifc_path = spec['mpi'].mpifc
+            # if we are using compiler wrappers on cray systems
+            # use those for mpi wrappers, b/c  spec['mpi'].mpicxx
+            # etc make return the spack compiler wrappers
+            # which can trip up mpi detection
+            if spec['mpi'].mpicc == spack_cc:
```

Comment:
If I am understanding this correctly, you want to use the MPI Compiler wrappers directly instead of querying the MPI compiler wrappers for information then passing that information to the non-mpi compilers.

You can do the following:

```
cfg.write(cmake_cache_option("ENABLE_MPI", True))
cfg.write(cmake_cache_option("ENABLE_FIND_MPI", False))
```

Then skip all of the MPI_*_COMPILER definitions.

Reply:
@white238  thanks for taking a look at this. 
It is not exactly what I'm doing, I think spec['mpi'].mpicc by default stores the Spack wrappers, so it would still need the re-assignment to the non-mpi compilers for Cray. 
Let me know if anything is unclear.

Reply:
@white238 @adityakavalur `spec['mpi'].mpicc` behaves differently on Cray systems from other systems. Cray systems don't have separate mpi compiler wrappers, and use the Cray compiler wrappers. So the compiler wrapping stack using MPI through Spack on a linux system is `MPI wrapper -> spack wrapper -> compiler`, and on Cray it is `spack wrapper -> Cray wrapper (handles MPI) -> compiler`.

As I understand it, `Axom`'s build system doesn't use the MPI compiler wrappers, it just uses them to detect the MPI libraries and then links MPI manually? Assuming that's true (it's common for cmake packages) then you should be able to use `spec['mpi'].mpicc` for the detection either way, as Spack will pass the argument through properly.

@adityakavalur, the issues that you've seen with this: does the issue manifest when building with Spack, or only when using the same hostconfig file to build outside of Spack?

Reply:
This is correct, Axom doesn't use the MPI compiler wrappers, except for information added to the regular compiler, unless you turn off ENABLE_FIND_MPI.  Then it assumes whatever you give to CMAKE_[C|CXX|Fortran]_COMPILER, can handle everything related to MPI.

Reply:
@becker33 : the issues I face are while building in Spack. The errors are essentially as you would expect, it is looking for MPI libraries and errors out when it cant find them (e.g. MPI_C_LIB_NAMES etc.)
@white238: Building with your modifications works as well, but as the 'find' is turned off, there is no way at build time to know if it is actually built correctly. I was attempting to point directly to the Cray compiler rather than Spack wrapper, since the former did not detect the libs.
Let me know if anything above is unclear.

Reply:
ENABLE_FIND_MPI is generally not recommended for reason it obfuscates what is really happening during the build, but is there for machines where the compiler wrappers don't report correctly to CMake's FindMPI module.

This seems to show a bug in the Spack compiler wrappers, as opposed to something that needs to be worked around in the Axom package, since what Axom is doing is the CMake standard way of handling MPI.

I am fine if this needs to be added temporarily while the cause of the problem is fixed but I will defer to @becker33

Reply:
As @becker33  described -- on cray systems, you really don't have a choice, there are no separate mpi compiler wrappers, and life is hard (at least different). I think this logic should only kick in for this case, so it seems ok to me.